lib/scholar/naive_bayes/categorical.ex

```diff
@@ -0,0 +1,556 @@
+    classes[Nx.argmax(jll, axis: 1)]
+  end
+
+  @doc """
+  Return log-probability estimates for the test vector `x` using `model`.
+
+  ## Examples
+
+      iex> x = Nx.iota({4, 3})
+      iex> y = Nx.tensor([1, 2, 0, 2])
+      iex> model = Scholar.NaiveBayes.Categorical.fit(x, y, num_classes: 3)
+      iex> Scholar.NaiveBayes.Categorical.predict_log_probability(model, Nx.tensor([[6, 2, 4], [8, 5, 9]]))
+      #Nx.Tensor<
+        f32[2][3]
+        [
+          [-0.8266787528991699, -1.5198254585266113, -1.0678410530090332],
+          [-1.272965431213379, -1.272965431213379, -0.8209810256958008]
+        ]
+      >
+  """
+
+  defn predict_log_probability(%__MODULE__{} = model, x) do
+    check_dim(x, Nx.axis_size(model.feature_count, 1))
+    jll = joint_log_likelihood(model, x)
+
+    log_proba_x =
+      jll
+      |> Nx.logsumexp(axes: [1])
+      |> Nx.reshape({Nx.axis_size(jll, 0), 1})
+
+    jll - log_proba_x
+  end
+
+  @doc """
+  Return probability estimates for the test vector `x` using `model`.
+
+  ## Examples
+
+      iex> x = Nx.iota({4, 3})
+      iex> y = Nx.tensor([1, 2, 0, 2])
+      iex> model = Scholar.NaiveBayes.Categorical.fit(x, y, num_classes: 3)
+      iex> Scholar.NaiveBayes.Categorical.predict_probability(model, Nx.tensor([[6, 2, 4], [8, 5, 9]]))
+      #Nx.Tensor<
+        f32[2][3]
+        [
+          [0.43749991059303284, 0.21875005960464478, 0.34374985098838806],
+          [0.28000006079673767, 0.28000006079673767, 0.4399997889995575]
+        ]
+      >
+  """
+
+  defn predict_probability(%__MODULE__{} = model, x) do
+    Nx.exp(predict_log_probability(model, x))
+  end
+
+  @doc """
+  Return joint log probability estimates for the test vector `x` using `model`.
+
+  ## Examples
+
+      iex> x = Nx.iota({4, 3})
+      iex> y = Nx.tensor([1, 2, 0, 2])
+      iex> model = Scholar.NaiveBayes.Categorical.fit(x, y, num_classes: 3)
+      iex> Scholar.NaiveBayes.Categorical.predict_joint_log_probability(model, Nx.tensor([[6, 2, 4], [8, 5, 9]]))
+      #Nx.Tensor<
+        f32[2][3]
+        [
+          [-8.140898704528809, -8.83404541015625, -8.382061004638672],
+          [-8.83404541015625, -8.83404541015625, -8.382061004638672]
+        ]
+      >
+  """
+
+  defn predict_joint_log_probability(%__MODULE__{} = model, x) do
+    check_dim(x, Nx.axis_size(model.feature_count, 1))
+    joint_log_likelihood(model, x)
+  end
+
+  defnp check_dim(x, dim) do
+    num_features = Nx.axis_size(x, 1)
+
+    if num_features != dim do
+      raise ArgumentError,
+            """
+            expected x to have same second dimension as data used for fitting model, \
+            got: #{num_features} for x and #{dim} for training data\
+            """
+    end
+  end
+
+  defnp joint_log_likelihood(
+          %__MODULE__{
+            feature_log_probability: feature_log_probability,
+            class_log_priors: class_log_priors
+          },
+          x
+        ) do
+    {_, _, _, jll} =
```

Comment:
As you don't use the first three variables, you can wrap them in a tuple, and then use more convenient pattern match `{_, jil}`